cg.go

```diff
@@ -0,0 +1,162 @@
+package optimize
+
+import (
+	"math"
+
+	"github.com/gonum/floats"
+)
+
+const (
+	iterationRestartFactor = 6
+	angleRestartTreshold   = 0.9
+)
+
+type CGVariant interface {
+	Init(loc Location)
+	Beta(gradient, gradPrev, dirPrev []float64) float64
+}
+
+type CG struct {
+	LinesearchMethod LinesearchMethod
+	Variant          CGVariant
+	InitialStep      StepSizer
+
+	IterationRestartFactor float64
+	AngleRestartTreshold   float64
+
+	restartAfter    int
+	iterFromRestart int
+
+	linesearch *Linesearch
+
+	gradPrev []float64
+	dirPrev  []float64
+}
+
+func (cg *CG) Init(loc Location, funcInfo *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	if cg.LinesearchMethod == nil {
+		cg.LinesearchMethod = &Bisection{GradConst: 0.1}
+	}
+	if cg.Variant == nil {
+		cg.Variant = &HestenesStiefel{}
+	}
+	if cg.InitialStep == nil {
+		cg.InitialStep = &FirstOrderStepSize{}
+	}
+
+	if cg.IterationRestartFactor == 0 {
+		cg.IterationRestartFactor = iterationRestartFactor
+	}
+	if cg.AngleRestartTreshold == 0 {
+		cg.AngleRestartTreshold = angleRestartTreshold
+	}
+
+	if cg.linesearch == nil {
+		cg.linesearch = &Linesearch{}
+	}
+	cg.linesearch.Method = cg.LinesearchMethod
+	cg.linesearch.NextDirectioner = cg
+
+	return cg.linesearch.Init(loc, funcInfo, xNext)
+}
+
+func (cg *CG) Iterate(loc Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return cg.linesearch.Iterate(loc, xNext)
+}
+
+func (cg *CG) InitDirection(loc Location, direction []float64) (stepSize float64) {
+	dim := len(loc.X)
+
+	cg.restartAfter = int(math.Ceil(cg.IterationRestartFactor * float64(dim)))
+	cg.iterFromRestart = 0
+	cg.gradPrev = resize(cg.gradPrev, dim)
+	cg.dirPrev = resize(cg.dirPrev, dim)
+
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	copy(cg.dirPrev, direction)
+	copy(cg.gradPrev, loc.Gradient)
+
+	cg.Variant.Init(loc)
+	return cg.InitialStep.Init(loc, direction)
+}
+
+func (cg *CG) NextDirection(loc Location, direction []float64) (stepSize float64) {
+	copy(direction, loc.Gradient)
+	floats.Scale(-1, direction)
+
+	var β float64
+	cg.iterFromRestart++
+	restart := cg.iterFromRestart%cg.restartAfter == 0
+	if !restart {
+		β = cg.Variant.Beta(loc.Gradient, cg.gradPrev, cg.dirPrev)
+	}
+	if β != 0 {
+		floats.AddScaled(direction, β, cg.dirPrev)
```

Comment:
Could you add a quick high-level comment here.

// If beta == 0, it means that it is time to restart (because of iteration number or the CGVariant). When we restart, ...